### ISSUE TITLE
Feature request: Continue automatic logging for a task that was stopped

### ISSUE THREAD

**Issue by michaelitvin** [allegroai/trains#160]
_(from [stackoverflow](https://stackoverflow.com/questions/62654203/trains-can-i-reset-the-status-of-a-task-from-aborted-back-to-running/62656158#62656158))_

Training can stop mid-way for various reasons, intended or unintended.
It would be nice to be able to continue automatic logging once training is restarted.

Something like:
```
task = Task.get_task('...')
task.continue_auto_logging()
```

**Comment by bmartinn** [comment]
Thanks @michaelitvin 
In order to differentiate from "automagic" logging and Task interfacing, 
I suggest we stick with Task.init to continue previous runs, and keep Task.get_task for querying interface

Maybe something like:
```python
task = Task.init('example', 'my_previous_run', reuse_last_task_id='aabbcc123')
```
My only concern is whether we should make sure the name/project match the previous run...
WDYT?

**Comment by michaelitvin** [comment]
Calling this `init()` is imho a bit confusing - the task already exists and it was inited before.
But I see your logic as well, so not sure what is the right thing here.

**Comment by bmartinn** [comment]
Hi @michaelitvin 
Good news, you can now continue previously executed experiments :rocket:
With the new `trains` release (0.16) you can now do:
```python
task = Task.init('examples', 'continue-run example', continue_last_task=True)
```
Which will continue the training of the previously executed experiment called 'continue-run example'.
Notice that if you pass a specific task_id to `continue_last_task`, you can continue the execution of any task in the system `continue_last_task='aabbcc112233'` , see details [here](https://allegro.ai/docs/task.html#trains.task.Task.init).

**Comment by michaelitvin** [comment]
Awesome! :)

**Comment by michaelitvin** [comment]
What happens if I continue_last_task from a new process while the original task is still running? Would it work? Would the logs be separated?

(Reminding my scenario with training and evaluation running in separate processes)

**Comment by bmartinn** [comment]
> What happens if I continue_last_task from a new process while the original task is still running? Would it work? Would the logs be separated?

A bit of implementation background:
If `continue_last_task` is True, the experiment that will be continued is the last experiment executed on the same machine with matching the project/experiment name. This means that if another machine executed an experiment with the same project/experiment name, you will not automatically continue that experiment.

Now regrading the specific question, which process are you "continuing" ? 
Running evaluation is not considered "continuing" the experiment. Continuing an experiment means that any new iteration reported, will be added to the last previously reported iteration (i.e. if the first run ended at iteration 1000, the new run automatically starts reporting at iteration 1001).

In your train/evaluate scenario, I don't think you need the "continue" feature.
You can always report to the same experiment from multiple machines, as long as you know the Task ID (or the task/project name is a unique combination).
Basically the evaluation process can do:
```python
running_task = Task.get_task(task_id='aabbcc')
last_iter = running_task.get_last_iteration()
running_task.get_logger().report_scalar('validation', 'accuracy', value=0.99, iteration=last_iter)
```

WDYT?

**Comment by michaelitvin** [comment]
Sounds good, the only thing I'm missing here is automatic capture of the console log of evaluation.

**Comment by bmartinn** [comment]
It sounds like these two issues (+ #181) should be connected , as the solution is quite similar.
What do you think?

**Comment by michaelitvin** [comment]
I don't know, but it would be nice to keep the logs separated - if they come from 2 parallel processes this can create a mess.

**Comment by KaidiGuo** [comment]
> Hi @michaelitvin Good news, you can now continue previously executed experiments rocket With the new `trains` release (0.16) you can now do:
> 
> ```python
> task = Task.init('examples', 'continue-run example', continue_last_task=True)
> ```
> 
> Which will continue the training of the previously executed experiment called 'continue-run example'. Notice that if you pass a specific task_id to `continue_last_task`, you can continue the execution of any task in the system `continue_last_task='aabbcc112233'` , see details [here](https://allegro.ai/docs/task.html#trains.task.Task.init).

Hi, I tried this method, it does continue to push the console log which is good, but it brings another problem. In my code, if I stopped the training and resume, I have a mechanism to restore to which epoch I have trained, in this case, after I re-init the task with continue_last_task to True, if I resume from iteration =10 and pass in 10 to my logger: `clearml_logger.report_scalar("frr", label, frr, iteration)` the actual iteration shown up on the plot will be 10+the the last iteration where I stopped. 
![Screenshot from 2022-11-03 16-56-13](https://user-images.githubusercontent.com/16896255/199833670-fb79ea99-01d2-4ede-ac74-09a2a3cdb220.png)
Like in this screenshot, the iteration value is jumping from iteration 37 to 75, because the actual epoch 38 became 37+38.
Is there a way to not do this iteration add up, but set it to actually be 38? Or is there any other way to continue the console log after I restart my code? I am currently doing a bypass which is using logger.report_text() to push my important log as text, but it would be nice if I could continue the automatic logging.

**Comment by erezalg** [comment]
Hi @KaidiGuo,

How are you getting the iteration parameter you use? Do you input there 10? Theoretically I don't think it should add, you actually might only need to provide it 0 and it should work internally in the SDK (IE, knowing that iteration 0 is actually 38)

**Comment by erezalg** [comment]
@KaidiGuo,

Just re-read docstrings of the Task.init() if you provide continue_last_task=0 it should not have this offset.
Let me know if it worked for you!

**Comment by KaidiGuo** [comment]
> @KaidiGuo,
> 
> Just re-read docstrings of the Task.init() if you provide continue_last_task=0 it should not have this offset. Let me know if it worked for you!

Perfect! This is exactly what I am looking for, now my loggers are working all right. But I thought continue_last_task is a bool type parameter, is it not? I see that you can actually pass in task ID in case you have duplicated task name. I couldn't find the corresponding doc...

Let me sum up quickly for other's sake:
When I want to resume a training process using the same task, ie. continue logging more iterations on the existing scalars and pushing console logs.

1. If I call `Task.init(project_name=project, task_name=experiment)`  again, I end up with a new task page that has the same name (different ID).

2. If I call `Task.init(config_parameters.project_name, config_parameters.experiment_name, continue_last_task=True)`, my new logs got pushed as desired, but all my following iteration in the scalars got shifted with +N, where N = last trained epoch.

3.  If I call `Task.init(config_parameters.project_name, config_parameters.experiment_name, continue_last_task=0)`, it will reset the continue offset to 0, thus my following iteration in the scalars will not shift, but be the same as the value I pass in to my `clearml_logger.report_scalar("frr", label, frr, iteration)`.

My next question is, how to achieve 3 when you have multiple tasks share the same name (let's say you accidentally created two tasks with the same name). I tried with continue_last_task='some_task_id', if I do so, my logger will have the shift again. How can I use task ID to specify a task I want to continue at the same time set the offset to 0?

**Comment by erezalg** [comment]
Hi @KaidiGuo,

As far as I understand, setting continue_last_task will continue the task with the same name that ran last and that's what the feature was. I'm not sure it's meant to continue arbitrary tasks.

Makes sense?

**Comment by patrickdasilva-clearml** [comment]
Hi everyone, we're closing this issue as stale since it's been created too long ago. If you think the issue is still relevant, please make sure the issue persists on a recent version of ClearML (ideally the latest) before re-opening the issue. We'll gladly look at it again.
